src/expr/parser.rs

```diff
@@ -1,58 +1,105 @@
-// Alex Ozdemir <aozdemir@hmc.edu> // <- Your name should replace this line!
+// Luis Viornery <lviornery@hmc.edu>
 // Starter code for HMC's MemorySafe, week 2
 //
 // The parser for an `Expr` (currently just produces the value fo the `Expr`)
 
+
+/*
+Workflow:
+start parsing additions => 
+start parsing multiplications => 
+start parsing literals => 
+finish parsing literals => 
+start parsing parens => 
+nested expr parsing => 
+finsish parsing parens => 
+finish parsing multiplication => 
+finish parsing additions
+*/
+
 use nom::digit;
 
 use std::str;
+use std::mem;
 use std::str::FromStr;
+use super::Expr;
+use super::BinOp;
 
-named!(parens<i64>, delimited!(
+//Make sure everything's return type is an expr
+
+named!(parens<Expr>,
+  delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
+named!(literal_expr<Expr>,
+  map!(
     map_res!(
-      digit,
-      str::from_utf8
-    ),
-    FromStr::from_str
+        map_res!(
+          digit,
+          str::from_utf8
+        ),
+        FromStr::from_str
+      ),
+    //This wraps the isize in an expr literal
+    Expr::Literal
+  )
+);
+
+named!(literal_expr_wrapper<Expr>,
+  alt!(
+    //If we uncomment these, we get sad segfaults - this is a whitespace implementation attempt
+    //preceded!(char!(' '),literal_expr_wrapper) |
+    //terminated!(literal_expr_wrapper, char!(' ')) |
```

Comment:
This is the problem.


Also, this strategy isn't quite enough. It couldn't parse:

```
( ( 5  + 6 )  / 6 )
```

because of spaces not next to literals.

To make it work you basically need to prefix all symbols: literals, parens, and operators with potential spaces. `preceeded!(multispace,...)` is useful. Since you prefix all symbols (and handle trailing spaces on the line somehow, no left-recursion happens.